### PR TITLE
adding aistore datapipe

### PR DESCRIPTION
**Pull request by gaikwadabhishek**
Fixes #{[517](https://github.com/pytorch/data/issues/517)}

### Changes
- Added `aisio.py` (Iterable Datapipe for AIStore backends)
- Added unit tests in ~~`test/test_local_io.py`~~ `test/test_aistore.py`
- Added GitHub action for running AIStore in ~~`CI.yml`~~ `.github/aistore_ci.yml` workflow

### Questions to maintainers
- We are unsure about the documentation generated on PyTorch to refer it in `README.md`, so I have tentatively added a URL similar to s3io functions. (see ```torchdata/datapipes/iter/load/README.md```)

Signed-off-by: Abhishek Gaikwad <gaikwadabhishek1997@gmail.com>

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (gaikwadabhishek) [reply .github/workflows/ci.yml:74]: **Running AIStore as a Docker Container**
Generally, AIStore is a storage cluster that requires a volume (ability to mount using `-v` in docker). In Github workflow environments, when we try to mount the `dev/root` as a volume to the container we receive an error during the startup which says that `dev/root` doesn't have underlying block storage. We can probably overcome this by running it from `/` or `/tmp` but we at AIStore don't recommend this. Currently, we are running it as a local deployment from scratch as a [Github action](https://github.com/NVIDIA/aistore/blob/master/action.yml).
**Running AIStore in different pipeline**
We can run AIStore as a separate workflow and import it into the CI workflow to run the python tests. But then we need to have a reliable mechanism to synchronize the testing pipeline with the AIS cluster starting up (in a separate pipeline). The latter may take anywhere between tens of milliseconds to maybe 10-20s depending on the load of the machine we run in. I have tried to simulate the [torch/data pipeline of this PR](https://github.com/gaikwadabhishek/data/runs/7032258015). 
 
Let me know what you guys think, we can work on this.

Comment (ejguan) [reply .github/workflows/ci.yml:74]: The reason I want to have a separate workflow is we can execute tests for AIStore within that workflow without touching others. With separate workflow, I am able to disable the test easily using GHA whenever we find constant failing tests.

Comment (ejguan) [reply .github/workflows/ci.yml:74]: > We can run AIStore as a separate workflow and import it into the CI workflow to run the python tests.

You don't need to do this. Since you AIStore doesn't require AWSSDK installed, you can basically copy paste the most of parts from `ci.yml` by removing all AWS related steps like:
- https://github.com/pytorch/data/blob/75f31dc99dcc3159ee9a3df73d3862c76754823c/.github/workflows/ci.yml#L44-L49
- https://github.com/pytorch/data/blob/75f31dc99dcc3159ee9a3df73d3862c76754823c/.github/workflows/ci.yml#L54-L61
- https://github.com/pytorch/data/blob/75f31dc99dcc3159ee9a3df73d3862c76754823c/.github/workflows/ci.yml#L70-L71
- Remove BUILD_S3=1. https://github.com/pytorch/data/blob/75f31dc99dcc3159ee9a3df73d3862c76754823c/.github/workflows/ci.yml#L75-L76
- Modify the tests steps to only execute the tests for AIStore.

File: setup.py
Comment (ejguan) [inline setup.py:200]: Since `aistore` is an optional dependency, we don't need to add it here.

File: .github/workflows/ci.yml
Comment (gaikwadabhishek) [reply .github/workflows/ci.yml]: Okay, I have made the changes. Do you want to trigger this workflow on a specific label? Currently the condition is similar to the main `ci.yml` -
```if:
      ${{ github.repository_owner == 'pytorch' && (github.event.action != 'labeled' ||
      startsWith(github.event.label.name, 'ciflow')) }}```

**Review by NivekT — COMMENTED** [review]
This PR mostly looks good to me. I will let @ejguan comment on the proper CI setup.

Question: Are there AWS/GCP/Azure remote setups that your library currently test against for every build? We might want to something similar here for different DataPipes (not just `aistore`).

**Comment by msaroufim** [comment]
> This PR mostly looks good to me. I will let @ejguan comment on the proper CI setup.
> 
> Question: Are there AWS/GCP/Azure remote setups that your library currently test against for every build? We might want to something similar here for different DataPipes (not just `aistore`).

Let's talk about this offline more @NivekT I was thinking of something similar for benchmarks where I can setup some machine configuration an instance type with some specific configuration for example HDD vs SSD and then run a script directly on it. There's a few options like `ray up`, cloudformation, ec2 launch, eks, teraform and I'm in the process of figuring out what's most convenient. In your case is the multi cloud requirement a must? In which case teraform may be the obvious thing to leverage more

File: .github/workflows/aistore_ci.yml
Comment (ejguan) [inline .github/workflows/aistore_ci.yml:20]: Like you said, no need to run the workflow for a specific label

File: .github/workflows/aistore_ci.yml
Comment (ejguan) [inline .github/workflows/aistore_ci.yml:53]: nit: you can remove this line for $GITHUB_PATH

File: test/test_aistore.py
Comment (ejguan) [inline test/test_aistore.py:21]: Noob question because I am not familiar with this syntax. Should it be
`except (ImportError, ConnectionError):`?

File: test/test_aistore.py
Comment (ejguan) [inline test/test_aistore.py:45]: Would the client copy the tempfile to the bucket? IIRC, `tempfile` will be removed from the fs on the completion of the context.

File: torchdata/datapipes/iter/load/aisio.py
Comment (ejguan) [inline torchdata/datapipes/iter/load/aisio.py:50]: Do you mind change the order of these two arguments to the same order as the signature.

File: torchdata/datapipes/iter/load/aisio.py
Comment (ejguan) [inline torchdata/datapipes/iter/load/aisio.py:97]: Ditto

**Review by ejguan — COMMENTED** [review]
Thank you for your contribution. Generally looks good to me. And, could you please sign the CLA before we can merge the PR?

File: test/test_aistore.py
Comment (gaikwadabhishek) [reply test/test_aistore.py:45]: Yes, the file created will be added to the AIS Bucket. Tempfile are deleted as soon as they are closed. In `tearDown()` we are destroying the bucket and all the objects within.

File: test/test_aistore.py
Comment (gaikwadabhishek) [reply test/test_aistore.py]: fixed

**Comment by gaikwadabhishek** [comment]
> Thank you for your contribution. Generally looks good to me. And, could you please sign the CLA before we can merge the PR?

Hi @ejguan @NivekT 
I am still waiting for my FB CLA to be approved. I work for NVIDIA and I've been added to the corporate CLA by my employer. We are waiting for the changes to reflect here. We have also mailed to cla@fb.com regarding this and are waiting for their reply.

**Comment by gaikwadabhishek** [comment]
@ejguan fixed the `mypi.ini` configuration to ignore the missing import of `aistore.`. Can you please retrigger the workflow

**Comment by ejguan** [comment]
FYI: The failures from TorchVision CI are unrelated to your PR.

**Comment by msaroufim** [comment]
@gaikwadabhishek your CLA application should have been approved just retrigger CI with an empty commit and the Facebook CLA check should pass

**Comment by msaroufim** [comment]
@zpao as FYI looks like CLA Check is still red

**Comment by gaikwadabhishek** [comment]
Hi @ejguan @NivekT @msaroufim,
Thanks for your support. 
CLA is approved now. Can we merge this?

**Review by msaroufim — APPROVED** [review]

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**
